File: src/main/java/org/kite9/diagram/visualization/planarization/rhd/grouping/generators/AbstractMergeGenerator.java
package org.kite9.diagram.visualization.planarization.rhd.grouping.generators;

import org.kite9.diagram.adl.Container;
import org.kite9.diagram.position.Direction;
import org.kite9.diagram.visualization.planarization.rhd.GroupPhase;
import org.kite9.diagram.visualization.planarization.rhd.GroupPhase.Group;
import org.kite9.diagram.visualization.planarization.rhd.grouping.basic.BasicMergeState;
import org.kite9.framework.logging.Kite9Log;
import org.kite9.framework.logging.Logable;

public abstract class AbstractMergeGenerator implements MergeGenerator, Logable  {

	protected GroupPhase gp;
	protected BasicMergeState ms;
	protected GeneratorBasedGroupingStrategy grouper;
	protected Kite9Log log = new Kite9Log(this);

	public AbstractMergeGenerator(GroupPhase gp, BasicMergeState ms, GeneratorBasedGroupingStrategy grouper) {
		super();
		this.gp = gp;
		this.ms = ms;
		this.grouper = grouper;
	}
	
	public void addMergeOption(Group g1, Group g2, Group alignedGroup, Direction alignedSide) {
		grouper.addMergeOption(g1, g2, alignedGroup, alignedSide, getMyBestPriority(), ms);
	}

	@Override
	public String getPrefix() {
		return "MG  ";
	}

	@Override
	public boolean isLoggingEnabled() {
		return true;
	}
	
	protected abstract int getMyBestPriority();
	
	@Override
	public void containerIsLive(Container c) {
	}
	
	protected abstract String getCode();
}

File: src/main/kotlin/org/kite9/diagram/visualization/planarization/rhd/grouping/generators/AbstractMergeGenerator.kt
package org.kite9.diagram.visualization.planarization.rhd.grouping.generators

import org.kite9.diagram.logging.Kite9Log
import org.kite9.diagram.logging.Logable
import org.kite9.diagram.model.Container
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.visualization.planarization.rhd.GroupPhase
import org.kite9.diagram.visualization.planarization.rhd.grouping.basic.group.Group
import org.kite9.diagram.visualization.planarization.rhd.grouping.basic.merge.BasicMergeState

abstract class AbstractMergeGenerator(
    val gp: GroupPhase,
    val ms: BasicMergeState,
    val grouper: GeneratorBasedGroupingStrategy
) : MergeGenerator, Logable {

    @JvmField
	protected var log = Kite9Log.instance(this)

    open fun addMergeOption(
        g1: Group,
        g2: Group,
        alignedGroup: Group?,
        alignedSide: Direction?
    ) {
        grouper.addMergeOption(g1, g2, alignedGroup, alignedSide, getMyBestPriority(), ms)
    }

    override val prefix: String
        get() = "MG  "

    override val isLoggingEnabled: Boolean
        get() = true

    protected abstract fun getMyBestPriority(): Int

    override fun containerIsLive(c: Container) {}

    abstract fun getCode(): String

}